### DISCUSSION TITLE
Getting pdfa

### DISCUSSION THREAD

**Discussion by roiDaniela** [DES-Lab/AALpy#34] [Q&A] [answered]
Hello,

I'm trying to get this prefix tree and learn it with alergia algorithm
![image](https://user-images.githubusercontent.com/11479074/177587557-4ad9a72c-3e8b-4400-8188-ae769b5af166.png)

I tried to build it using mdp but can not find how to set the states as accepting states
thanks
   
     `len_input = 4
    states = []
    for i in range(13):
        states.append(MdpState(f'q{i}'))

    # 0
    states[0].transitions['a'].append((states[1], 0.25))
    states[0].transitions['a'].append((states[0], 0.75))

    states[0].transitions['b'].append((states[4], 0.25))
    states[0].transitions['b'].append((states[0], 0.75))

    states[0].transitions['c'].append((states[7], 0.25))
    states[0].transitions['c'].append((states[0], 0.75))

    states[0].transitions['d'].append((states[10], 0.25))
    states[0].transitions['d'].append((states[0], 0.75))

    # 1
    states[1].transitions['a'].append((states[2], 0.25))
    states[1].transitions['a'].append((states[1], 0.75))

    states[1].transitions['b'].append((states[1], 0.2))
    states[1].transitions['b'].append((states[1], 0.8))

    states[1].transitions['c'].append((states[1], 0.2))
    states[1].transitions['c'].append((states[1], 0.8))

    states[1].transitions['d'].append((states[1], 0.2))
    states[1].transitions['d'].append((states[1], 0.8))

    # 2
    states[2].transitions['a'].append((states[3], 0.25))
    states[2].transitions['a'].append((states[2], 0.75))

    states[2].transitions['b'].append((states[2], 0.2))
    states[2].transitions['b'].append((states[2], 0.8))

    states[2].transitions['c'].append((states[2], 0.2))
    states[2].transitions['c'].append((states[2], 0.8))

    states[2].transitions['d'].append((states[2], 0.2))
    states[2].transitions['d'].append((states[2], 0.8))

    # 3
    states[3].transitions['a'].append((states[3], 0.25))
    states[3].transitions['a'].append((states[3], 0.75))

    states[3].transitions['b'].append((states[3], 0.2))
    states[3].transitions['b'].append((states[3], 0.8))

    states[3].transitions['c'].append((states[3], 0.2))
    states[3].transitions['c'].append((states[3], 0.8))

    states[3].transitions['d'].append((states[3], 0.2))
    states[3].transitions['d'].append((states[3], 0.8))

    # 4
    states[4].transitions['a'].append((states[4], 0.2))
    states[4].transitions['a'].append((states[4], 0.8))

    states[4].transitions['b'].append((states[5], 0.25))
    states[4].transitions['b'].append((states[4], 0.75))

    states[4].transitions['c'].append((states[4], 0.2))
    states[4].transitions['c'].append((states[4], 0.8))

    states[4].transitions['d'].append((states[4], 0.2))
    states[4].transitions['d'].append((states[4], 0.8))
    
    # 5
    states[5].transitions['a'].append((states[5], 0.2))
    states[5].transitions['a'].append((states[5], 0.8))

    states[5].transitions['b'].append((states[6], 0.25))
    states[5].transitions['b'].append((states[5], 0.75))

    states[5].transitions['c'].append((states[5], 0.2))
    states[5].transitions['c'].append((states[5], 0.8))

    states[5].transitions['d'].append((states[5], 0.2))
    states[5].transitions['d'].append((states[5], 0.8))
    
    # 6
    states[6].transitions['a'].append((states[6], 0.2))
    states[6].transitions['a'].append((states[6], 0.8))

    states[6].transitions['b'].append((states[6], 0.25))
    states[6].transitions['b'].append((states[6], 0.75))

    states[6].transitions['c'].append((states[6], 0.2))
    states[6].transitions['c'].append((states[6], 0.8))

    states[6].transitions['d'].append((states[6], 0.2))
    states[6].transitions['d'].append((states[6], 0.8))
    
    # 7
    states[7].transitions['a'].append((states[7], 0.2))
    states[7].transitions['a'].append((states[7], 0.8))

    states[7].transitions['b'].append((states[8], 0.25))
    states[7].transitions['b'].append((states[7], 0.75))

    states[7].transitions['c'].append((states[7], 0.2))
    states[7].transitions['c'].append((states[7], 0.8))

    states[7].transitions['d'].append((states[7], 0.2))
    states[7].transitions['d'].append((states[7], 0.8))
    
    # 8
    states[8].transitions['a'].append((states[8], 0.2))
    states[8].transitions['a'].append((states[8], 0.8))

    states[8].transitions['b'].append((states[9], 0.25))
    states[8].transitions['b'].append((states[8], 0.75))

    states[8].transitions['c'].append((states[8], 0.2))
    states[8].transitions['c'].append((states[8], 0.8))

    states[8].transitions['d'].append((states[8], 0.2))
    states[8].transitions['d'].append((states[8], 0.8))

    # 9
    states[9].transitions['a'].append((states[9], 0.2))
    states[9].transitions['a'].append((states[9], 0.8))

    states[9].transitions['b'].append((states[9], 0.25))
    states[9].transitions['b'].append((states[9], 0.75))

    states[9].transitions['c'].append((states[9], 0.2))
    states[9].transitions['c'].append((states[9], 0.8))

    states[9].transitions['d'].append((states[9], 0.2))
    states[9].transitions['d'].append((states[9], 0.8))
    
    # 10
    states[10].transitions['a'].append((states[10], 0.2))
    states[10].transitions['a'].append((states[10], 0.8))

    states[10].transitions['b'].append((states[11], 0.25))
    states[10].transitions['b'].append((states[10], 0.75))

    states[10].transitions['c'].append((states[10], 0.2))
    states[10].transitions['c'].append((states[10], 0.8))

    states[10].transitions['d'].append((states[10], 0.2))
    states[10].transitions['d'].append((states[10], 0.8))
    
    # 11
    states[11].transitions['a'].append((states[11], 0.2))
    states[11].transitions['a'].append((states[11], 0.8))

    states[11].transitions['b'].append((states[11], 0.2))
    states[11].transitions['b'].append((states[11], 0.8))

    states[11].transitions['c'].append((states[12], 0.25))
    states[11].transitions['c'].append((states[11], 0.75))

    states[11].transitions['d'].append((states[11], 0.2))
    states[11].transitions['d'].append((states[11], 0.8))
    
    # 12
    states[12].transitions['a'].append((states[12], 0.2))
    states[12].transitions['a'].append((states[12], 0.8))

    states[12].transitions['b'].append((states[12], 0.2))
    states[12].transitions['b'].append((states[12], 0.8))

    states[12].transitions['c'].append((states[12], 0.25))
    states[12].transitions['c'].append((states[12], 0.75))

    states[12].transitions['d'].append((states[12], 0.2))
    states[12].transitions['d'].append((states[12], 0.8))
    
    return Mdp(states[0], states), list(range(len_input))`

**Comment by emuskardin** [comment]
Hi,

What I would do, is set the output to True for accepting and False for rejecting states.
```
MdpState(f'q{i}', output=True)
or
states[12].output = True
```

Let me know if that serves the purpose :)

**Reply by roiDaniela** [reply]
Thanks @emuskardin 

I got this for the original

![image](https://user-images.githubusercontent.com/11479074/177591767-059bd393-9a55-4b99-9d69-9d982f16a5e4.png)

and for the learned automaton
![image](https://user-images.githubusercontent.com/11479074/177591938-e6a1fee1-1265-4cf0-8afe-3415c89eaf6f.png)

is this ok?
the learned automaton is actually different

thanks
code is going like this:

```
    states = []
    for i in range(13):
        # curr_output = state_outputs.pop(0) if state_outputs else random.choice(outputs)
        if i == 3 or i == 6 or i == 9 or i == 12:
            states.append(MdpState(f'q{i}', output=True))
        else:
            states.append(MdpState(f'q{i}', output=False))

    # 0
    states[0].transitions['a'].append((states[1], 0.25))
    states[0].transitions['a'].append((states[0], 0.75))

    states[0].transitions['b'].append((states[4], 0.25))
    states[0].transitions['b'].append((states[0], 0.75))

    states[0].transitions['c'].append((states[7], 0.25))
    states[0].transitions['c'].append((states[0], 0.75))

    states[0].transitions['d'].append((states[10], 0.25))
    states[0].transitions['d'].append((states[0], 0.75))

    # 1
    states[1].transitions['a'].append((states[2], 0.25))
    states[1].transitions['a'].append((states[1], 0.75))

    states[1].transitions['b'].append((states[1], 0.2))
    states[1].transitions['b'].append((states[1], 0.8))

    states[1].transitions['c'].append((states[1], 0.2))
    states[1].transitions['c'].append((states[1], 0.8))

    states[1].transitions['d'].append((states[1], 0.2))
    states[1].transitions['d'].append((states[1], 0.8))

    # 2
    states[2].transitions['a'].append((states[3], 0.25))
    states[2].transitions['a'].append((states[2], 0.75))

    states[2].transitions['b'].append((states[2], 0.2))
    states[2].transitions['b'].append((states[2], 0.8))

    states[2].transitions['c'].append((states[2], 0.2))
    states[2].transitions['c'].append((states[2], 0.8))

    states[2].transitions['d'].append((states[2], 0.2))
    states[2].transitions['d'].append((states[2], 0.8))

    # 3
    states[3].transitions['a'].append((states[3], 0.25))
    states[3].transitions['a'].append((states[3], 0.75))

    states[3].transitions['b'].append((states[3], 0.2))
    states[3].transitions['b'].append((states[3], 0.8))

    states[3].transitions['c'].append((states[3], 0.2))
    states[3].transitions['c'].append((states[3], 0.8))

    states[3].transitions['d'].append((states[3], 0.2))
    states[3].transitions['d'].append((states[3], 0.8))

    # 4
    states[4].transitions['a'].append((states[4], 0.2))
    states[4].transitions['a'].append((states[4], 0.8))

    states[4].transitions['b'].append((states[5], 0.25))
    states[4].transitions['b'].append((states[4], 0.75))

    states[4].transitions['c'].append((states[4], 0.2))
    states[4].transitions['c'].append((states[4], 0.8))

    states[4].transitions['d'].append((states[4], 0.2))
    states[4].transitions['d'].append((states[4], 0.8))

    # 5
    states[5].transitions['a'].append((states[5], 0.2))
    states[5].transitions['a'].append((states[5], 0.8))

    states[5].transitions['b'].append((states[6], 0.25))
    states[5].transitions['b'].append((states[5], 0.75))

    states[5].transitions['c'].append((states[5], 0.2))
    states[5].transitions['c'].append((states[5], 0.8))

    states[5].transitions['d'].append((states[5], 0.2))
    states[5].transitions['d'].append((states[5], 0.8))

    # 6
    states[6].transitions['a'].append((states[6], 0.2))
    states[6].transitions['a'].append((states[6], 0.8))

    states[6].transitions['b'].append((states[6], 0.25))
    states[6].transitions['b'].append((states[6], 0.75))

    states[6].transitions['c'].append((states[6], 0.2))
    states[6].transitions['c'].append((states[6], 0.8))

    states[6].transitions['d'].append((states[6], 0.2))
    states[6].transitions['d'].append((states[6], 0.8))

    # 7
    states[7].transitions['a'].append((states[7], 0.2))
    states[7].transitions['a'].append((states[7], 0.8))

    states[7].transitions['b'].append((states[8], 0.25))
    states[7].transitions['b'].append((states[7], 0.75))

    states[7].transitions['c'].append((states[7], 0.2))
    states[7].transitions['c'].append((states[7], 0.8))

    states[7].transitions['d'].append((states[7], 0.2))
    states[7].transitions['d'].append((states[7], 0.8))

    # 8
    states[8].transitions['a'].append((states[8], 0.2))
    states[8].transitions['a'].append((states[8], 0.8))

    states[8].transitions['b'].append((states[9], 0.25))
    states[8].transitions['b'].append((states[8], 0.75))

    states[8].transitions['c'].append((states[8], 0.2))
    states[8].transitions['c'].append((states[8], 0.8))

    states[8].transitions['d'].append((states[8], 0.2))
    states[8].transitions['d'].append((states[8], 0.8))

    # 9
    states[9].transitions['a'].append((states[9], 0.2))
    states[9].transitions['a'].append((states[9], 0.8))

    states[9].transitions['b'].append((states[9], 0.25))
    states[9].transitions['b'].append((states[9], 0.75))

    states[9].transitions['c'].append((states[9], 0.2))
    states[9].transitions['c'].append((states[9], 0.8))

    states[9].transitions['d'].append((states[9], 0.2))
    states[9].transitions['d'].append((states[9], 0.8))

    # 10
    states[10].transitions['a'].append((states[10], 0.2))
    states[10].transitions['a'].append((states[10], 0.8))

    states[10].transitions['b'].append((states[11], 0.25))
    states[10].transitions['b'].append((states[10], 0.75))

    states[10].transitions['c'].append((states[10], 0.2))
    states[10].transitions['c'].append((states[10], 0.8))

    states[10].transitions['d'].append((states[10], 0.2))
    states[10].transitions['d'].append((states[10], 0.8))

    # 11
    states[11].transitions['a'].append((states[11], 0.2))
    states[11].transitions['a'].append((states[11], 0.8))

    states[11].transitions['b'].append((states[11], 0.2))
    states[11].transitions['b'].append((states[11], 0.8))

    states[11].transitions['c'].append((states[12], 0.25))
    states[11].transitions['c'].append((states[11], 0.75))

    states[11].transitions['d'].append((states[11], 0.2))
    states[11].transitions['d'].append((states[11], 0.8))

    # 12
    states[12].transitions['a'].append((states[12], 0.2))
    states[12].transitions['a'].append((states[12], 0.8))

    states[12].transitions['b'].append((states[12], 0.2))
    states[12].transitions['b'].append((states[12], 0.8))

    states[12].transitions['c'].append((states[12], 0.25))
    states[12].transitions['c'].append((states[12], 0.75))

    states[12].transitions['d'].append((states[12], 0.2))
    states[12].transitions['d'].append((states[12], 0.8))

    return Mdp(states[0], states)#, list(range(len_input))
```

**Reply by emuskardin** [reply]
Learned automaton will for sure be different, as the automaton you create can be minimized.

Can you post the data generation and call to Alergia as well, then I can also run it

**Reply by roiDaniela** [reply]
In file Examples.py I replaced 
in the function

`def alergia_mdp_example():
`

`    mdp = generate_random_mdp(5, 2, 5)
`
with
`    mdp = my_generate_random_mdp()
`

which is

```
def my_generate_random_mdp():
    states = []
    for i in range(13):
        # curr_output = state_outputs.pop(0) if state_outputs else random.choice(outputs)
        if i == 3 or i == 6 or i == 9 or i == 12:
            states.append(MdpState(f'q{i}', output=True))
        else:
            states.append(MdpState(f'q{i}', output=False))

    # 0
    states[0].transitions['a'].append((states[1], 0.25))
    states[0].transitions['a'].append((states[0], 0.75))

    states[0].transitions['b'].append((states[4], 0.25))
    states[0].transitions['b'].append((states[0], 0.75))

    states[0].transitions['c'].append((states[7], 0.25))
    states[0].transitions['c'].append((states[0], 0.75))

    states[0].transitions['d'].append((states[10], 0.25))
    states[0].transitions['d'].append((states[0], 0.75))

    # 1
    states[1].transitions['a'].append((states[2], 0.25))
    states[1].transitions['a'].append((states[1], 0.75))

    states[1].transitions['b'].append((states[1], 0.2))
    states[1].transitions['b'].append((states[1], 0.8))

    states[1].transitions['c'].append((states[1], 0.2))
    states[1].transitions['c'].append((states[1], 0.8))

    states[1].transitions['d'].append((states[1], 0.2))
    states[1].transitions['d'].append((states[1], 0.8))

    # 2
    states[2].transitions['a'].append((states[3], 0.25))
    states[2].transitions['a'].append((states[2], 0.75))

    states[2].transitions['b'].append((states[2], 0.2))
    states[2].transitions['b'].append((states[2], 0.8))

    states[2].transitions['c'].append((states[2], 0.2))
    states[2].transitions['c'].append((states[2], 0.8))

    states[2].transitions['d'].append((states[2], 0.2))
    states[2].transitions['d'].append((states[2], 0.8))

    # 3
    states[3].transitions['a'].append((states[3], 0.25))
    states[3].transitions['a'].append((states[3], 0.75))

    states[3].transitions['b'].append((states[3], 0.2))
    states[3].transitions['b'].append((states[3], 0.8))

    states[3].transitions['c'].append((states[3], 0.2))
    states[3].transitions['c'].append((states[3], 0.8))

    states[3].transitions['d'].append((states[3], 0.2))
    states[3].transitions['d'].append((states[3], 0.8))

    # 4
    states[4].transitions['a'].append((states[4], 0.2))
    states[4].transitions['a'].append((states[4], 0.8))

    states[4].transitions['b'].append((states[5], 0.25))
    states[4].transitions['b'].append((states[4], 0.75))

    states[4].transitions['c'].append((states[4], 0.2))
    states[4].transitions['c'].append((states[4], 0.8))

    states[4].transitions['d'].append((states[4], 0.2))
    states[4].transitions['d'].append((states[4], 0.8))

    # 5
    states[5].transitions['a'].append((states[5], 0.2))
    states[5].transitions['a'].append((states[5], 0.8))

    states[5].transitions['b'].append((states[6], 0.25))
    states[5].transitions['b'].append((states[5], 0.75))

    states[5].transitions['c'].append((states[5], 0.2))
    states[5].transitions['c'].append((states[5], 0.8))

    states[5].transitions['d'].append((states[5], 0.2))
    states[5].transitions['d'].append((states[5], 0.8))

    # 6
    states[6].transitions['a'].append((states[6], 0.2))
    states[6].transitions['a'].append((states[6], 0.8))

    states[6].transitions['b'].append((states[6], 0.25))
    states[6].transitions['b'].append((states[6], 0.75))

    states[6].transitions['c'].append((states[6], 0.2))
    states[6].transitions['c'].append((states[6], 0.8))

    states[6].transitions['d'].append((states[6], 0.2))
    states[6].transitions['d'].append((states[6], 0.8))

    # 7
    states[7].transitions['a'].append((states[7], 0.2))
    states[7].transitions['a'].append((states[7], 0.8))

    states[7].transitions['b'].append((states[8], 0.25))
    states[7].transitions['b'].append((states[7], 0.75))

    states[7].transitions['c'].append((states[7], 0.2))
    states[7].transitions['c'].append((states[7], 0.8))

    states[7].transitions['d'].append((states[7], 0.2))
    states[7].transitions['d'].append((states[7], 0.8))

    # 8
    states[8].transitions['a'].append((states[8], 0.2))
    states[8].transitions['a'].append((states[8], 0.8))

    states[8].transitions['b'].append((states[9], 0.25))
    states[8].transitions['b'].append((states[8], 0.75))

    states[8].transitions['c'].append((states[8], 0.2))
    states[8].transitions['c'].append((states[8], 0.8))

    states[8].transitions['d'].append((states[8], 0.2))
    states[8].transitions['d'].append((states[8], 0.8))

    # 9
    states[9].transitions['a'].append((states[9], 0.2))
    states[9].transitions['a'].append((states[9], 0.8))

    states[9].transitions['b'].append((states[9], 0.25))
    states[9].transitions['b'].append((states[9], 0.75))

    states[9].transitions['c'].append((states[9], 0.2))
    states[9].transitions['c'].append((states[9], 0.8))

    states[9].transitions['d'].append((states[9], 0.2))
    states[9].transitions['d'].append((states[9], 0.8))

    # 10
    states[10].transitions['a'].append((states[10], 0.2))
    states[10].transitions['a'].append((states[10], 0.8))

    states[10].transitions['b'].append((states[11], 0.25))
    states[10].transitions['b'].append((states[10], 0.75))

    states[10].transitions['c'].append((states[10], 0.2))
    states[10].transitions['c'].append((states[10], 0.8))

    states[10].transitions['d'].append((states[10], 0.2))
    states[10].transitions['d'].append((states[10], 0.8))

    # 11
    states[11].transitions['a'].append((states[11], 0.2))
    states[11].transitions['a'].append((states[11], 0.8))

    states[11].transitions['b'].append((states[11], 0.2))
    states[11].transitions['b'].append((states[11], 0.8))

    states[11].transitions['c'].append((states[12], 0.25))
    states[11].transitions['c'].append((states[11], 0.75))

    states[11].transitions['d'].append((states[11], 0.2))
    states[11].transitions['d'].append((states[11], 0.8))

    # 12
    states[12].transitions['a'].append((states[12], 0.2))
    states[12].transitions['a'].append((states[12], 0.8))

    states[12].transitions['b'].append((states[12], 0.2))
    states[12].transitions['b'].append((states[12], 0.8))

    states[12].transitions['c'].append((states[12], 0.25))
    states[12].transitions['c'].append((states[12], 0.75))

    states[12].transitions['d'].append((states[12], 0.2))
    states[12].transitions['d'].append((states[12], 0.8))

    return Mdp(states[0], states)#, list(range(len_input))

```

**Comment by emuskardin** [comment, marked as answer]
I would guess  that the learned automaton is what is whitin ALERGIA's  capabilites when it comes to learning of non-deterministic MDPs. So you get a deterministic approximation of non-det MDP.

What I mean by non-deterministic:
```
    states[0].transitions['a'].append((states[1], 0.25))
    states[0].transitions['a'].append((states[0], 0.75))
```
From the same state, same input `a` leads to a same output `False`. Even tho it is a different state, ALERGIA cannot differentiate it based on the output

**Comment by emuskardin** [comment]
We wrote about that in the preliminaries of the https://arxiv.org/pdf/2206.11708.pdf

**Reply by roiDaniela** [reply]
Thanks, I thought it is the only way I can create probability prefix tree in the structure 
![image](https://user-images.githubusercontent.com/11479074/177606273-f42ea7bd-833c-470b-8cb8-facb06a9a6bc.png)

Is there is a way to create determenistic mdp using AALPY?

**Reply by emuskardin** [reply]
You have successfully created the structure on the image, but that one is by default non-deterministic.

To best of my knowledge, there is no algorithm that can learn non-deterministic MDPs. ALERGIA can approximate it, and that is the 2 state model that you get when you run the alg.

**Reply by roiDaniela** [reply]
Thanks @emuskardin

**Reply by roiDaniela** [reply]
@emuskardin just tp make things clear, Alergia can not learn a detemenistic automaton with probability? (meaning that action 'a' has 0.5 prob and action 'b' has 0.5 prob ...)

**Comment by emuskardin** [comment]
Alergia learns a deterministic MDPs correctly in a limit, but what you have here is a non-deterministic MDP where action `a` has 25% probability of going to a state with a label `False` and 75% chance to go to another state with the same label, creating the non-determinism. 

If you have a deterministic MDP, meaning that each all transitions from a state-input pair lead to states with a different output, ALERGIA will learn the correct MDP in a limit. If multiple state-input transitions lead to different states with the same output, like what you have, it will create a deterministic approximation.

**Reply by roiDaniela** [reply]
Thanks @emuskardin, The motivation for what I'm looking for is to reconstruct the automaton that was claimed fails to learn with alergia in the article - https://www.aaai.org/Papers/Symposia/Spring/1996/SS-96-05/SS96-05-006.pdf.
Do you know this article? I'm not sure the example I'm trying is good counterexample.

**Reply by emuskardin** [reply]
I am doing last preparations before holiday, so I cannot dedicate the time to read the full paper now.
Some things to note:
- In the paper they say that this is a DFA, not a probabilistic DFA. If you want to learn a DFA, then use a RPNI algorithm. It is not just "deterministic ALERGIA", the way it works is slightly different.
- ALERGIA is not the best in itself with sequential data where you have to "count", like what you have here... If you need 2 `a` inputs to get to `True` with a third `a` input , there is a high chance that ALERGIA will simply have 2 states, where the probability of going to `True` output is 33%

**Reply by roiDaniela** [reply]
thanks, according to (1), the paper is scecificly regarding Alergia and not rpni (they assume probability)

**Reply by emuskardin** [reply]
Ok, I was just being specific for that example as they say DFA.
Recent algorithm that performs seems to be mroe precise than ALERGIA can be found [here](https://www.researchgate.net/profile/Raphael-Reynouard/publication/358125919_Active_Learning_of_Markov_Decision_Processes_using_Baum-Welch_algorithm/links/61fa51421e98d168d7e6c536/Active-Learning-of-Markov-Decision-Processes-using-Baum-Welch-algorithm.pdf), though I cannot say any more about it since I have never implemented/used it.

**Reply by roiDaniela** [reply]
> * ALERGIA is not the best in itself with sequential data where you have to "count", like what you have here... If you need 2 `a` inputs to get to `True` with a third `a` input , there is a high chance that ALERGIA will simply have 2 states, where the probability of going to `True` output is 33%

Why "a" is not a distinguish sequence between "a" and "a-a"? Althogh they have same probability but compatible test also check the output and the output of "aa" and "aaa" is different?
Thanks
